nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -0,0 +1,135 @@
+            while status_data[0] != 1 and status_data[0] != 2: # FAT END POINT
+                time.sleep(0.001)
+                pass
+            # wait until the data can be send to the sender thread
+            # Set lock to true and put the data in the shared buffer
+            buffer[0] = store.return_data()
+            with lock_status: # FAT END POINT
+                if status_data[0] != 2:
+                    status_data[0] = 0
+        elif status_.Get_tag() == 1:
+            logger.info("Nest to TVB : receive end " + str(count))
+            count += 1
+        elif status_.Get_tag() == 2:
+            with lock_status:
+                status_data[0] = 2
+            logger.info("end simulation")
+            break
+        else:
+            raise Exception("bad mpi tag"+str(status_.Get_tag()))
+    logger.info('communication disconnect')
+    comm.Disconnect()
+    logger.info('end thread')
+    return
+
+
+def send(logger,analyse,status_data,buffer, comm):
+    '''
+    the sending part of the translator
+    :param logger: logger fro the thread
+    :param analyse ; analyse object to transform spikes to state variable
+    :param status_data: the status of the buffer (SHARED between thread)
+    :param buffer: the buffer which contains the data (SHARED between thread)
+    :return:
+    '''
+    count=0
+    status_ = MPI.Status()
+    while True: # FAT END POINT
+        # wait until the translator accept the connections
+        accept = False
+        logger.info("Nest to TVB : wait to send " )
+        while not accept:
+            req = comm.irecv(source=MPI.ANY_SOURCE,tag=MPI.ANY_TAG)
+            accept = req.wait(status_)
+        logger.info(" Nest to TVB : send data status : " +str(status_.Get_tag()))
+        if status_.Get_tag() == 0:
+            # send the rate when there ready
+            while status_data[0] != 0: # FAT END POINT
+                time.sleep(0.001)
+                pass
+            times,data=analyse.analyse(count,buffer[0])
+            with lock_status:
+                if status_data[0] != 2:
+                    status_data[0] = 1
+            logger.info("Nest to TVB : send data :"+str(np.sum(data)) )
+            # time of stating and ending step
+            comm.Send([times, MPI.DOUBLE], dest=status_.Get_source(), tag=0)
+            # send the size of the rate
+            size = np.array(int(data.shape[0]),dtype='i')
+            comm.Send([size,MPI.INT], dest=status_.Get_source(), tag=0)
+            # send the rates
+            comm.Send([data,MPI.DOUBLE], dest=status_.Get_source(), tag=0)
+        elif status_.Get_tag() == 1:
+            # disconnect when everything is ending
+            with lock_status:
+                status_data[0] = 2
+            break
+        else:
+            raise Exception("bad mpi tag"+str(status_.Get_tag()))
+        count+=1
+    logger.info('communication disconnect')
+    comm.Disconnect()
+    logger.info('end thread')
+    return
```

Comment:
Spurious return

Reply:
Old code of the threaded version. Will be changed.